### PR TITLE
Enable Deployer Whitelist

### PR DESCRIPTION
**Pull request by ben-chain**
## Description
Adds checks in the execution manager to only allow deployments if the deployer is whitelisted by, or arbitrary deployment is enabled in, the `OVM_DeployerWhitelist`.

### Fixes
- Fixes roadmap#254

## Contributing Agreement

- [x] I have read and understood the [Optimism Contributing Guide and Code of Conduct](https://github.com/ethereum-optimism/optimism-monorepo/blob/master/.github/CONTRIBUTING.md) and am following those guidelines in this pull request.

### CODE REVIEW

File: contracts/optimistic-ethereum/OVM/precompiles/OVM_DeployerWhitelist.sol [hunk 1]
```diff
@@ -7,6 +7,7 @@ import { Lib_Bytes32Utils } from "../../libraries/utils/Lib_Bytes32Utils.sol";
 /* Interface Imports */
 import { iOVM_DeployerWhitelist } from "../../iOVM/precompiles/iOVM_DeployerWhitelist.sol";
 import { iOVM_ExecutionManager } from "../../iOVM/execution/iOVM_ExecutionManager.sol";
+import { console } from "@nomiclabs/buidler/console.sol";
```
Comment (karlfloersch) [inline contracts/optimistic-ethereum/OVM/precompiles/OVM_DeployerWhitelist.sol]: Forgot to remove console log?

File: contracts/optimistic-ethereum/OVM/precompiles/OVM_DeployerWhitelist.sol [hunk 2]
```diff
@@ -178,6 +179,9 @@ contract OVM_DeployerWhitelist is iOVM_DeployerWhitelist {
             ovmExecutionManager.ovmSLOAD(KEY_INITIALIZED)
         );
 
+        console.log("initialized is: ");
+        console.logBool(initialized);
+
```
Comment (karlfloersch) [inline contracts/optimistic-ethereum/OVM/precompiles/OVM_DeployerWhitelist.sol]: Also here

File: contracts/test-helpers/Helper_TestRunner.sol [hunk 3]
```diff
@@ -14,13 +14,16 @@ contract Helper_TestRunner {
...
+        bool onlyValidateFlag;
     }
 
     function runSingleTestStep(
         TestStep memory _step
     )
         public
     {
+        console.log("onlyValidateFlag is:");
+        console.logBool(_step.onlyValidateFlag);
```
Comment (karlfloersch) [inline contracts/test-helpers/Helper_TestRunner.sol]: More logs but I guess this is in a test file so not key to remove if needed

**Review by karlfloersch — APPROVED** [review]
This looks great to me! I'm not suuuper familiar with the EM tooling but the changes you made at a glance make sense. I think the best thing to do before merging is just do a quick integration test using `up.sh` and disabling arbitrary contract deployment to at least observe that it all works. Of course even better than that would be adding a test to the `integration-tests` repo

**Comment by ben-chain** [comment]
Let's merge this into separate branch and integration test first so things appear working.

**Review by kfichter — APPROVED** [review]
LGTM

File: contracts/optimistic-ethereum/OVM/execution/OVM_ExecutionManager.sol [hunk 4]
```diff
@@ -844,7 +853,32 @@ contract OVM_ExecutionManager is iOVM_ExecutionManager, Lib_AddressResolver {
...
+     * Checks whether the given address is on the whitelst to ovmCREATE/ovmCREATE2, and reverts if not.
+     * @param _deployerAddress Address attempting to deploy a contract.
+     */
+    function _checkDeployerAllowed(
+        address _deployerAddress
+    )
+        internal
+    {
+        // From an OVM semanitcs perspectibe, this will appear the identical to
+        //  the deployer ovmCALLing the whitelist.  This is fine--in a sense, we are forcing them to.
```
Comment (ben-chain) [inline contracts/optimistic-ethereum/OVM/execution/OVM_ExecutionManager.sol:870]:
```suggestion
        // the deployer ovmCALLing the whitelist.  This is fine--in a sense, we are forcing them to.
```

**Pull request merged**